[SILOptimizer/stdlib] Emit failing _preconditions as compiler errors

JaSpa opened the pull request:
#### What's in this pull request?

This change allows failing `_precondition`s inside `@_transparent` functions to be emitted as errors.
#### What's not in this pull request?

Only the preconditions in inlined functions can be folded to a constant true/false value, so the only functions taking full leverage of this change are the range returning overloads where `Pos` is `Comparable` of `...` and `..<`.
#### Resolved bug number: [SR-707](https://bugs.swift.org/browse/SR-707)

<!-- If this pull request resolves any bugs from Swift bug tracker -->

---

<!-- This selection should only be completed by Swift admin -->

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI </summary>



The swift-ci is triggered by writing a comment on this PR addressed to the github user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

<!-- Thank you for your contribution to Swift! -->

JaSpa committed:
[SILOptimizer] Print user specified messages with `staticReport`

If the third argument to the `staticReport` builtin is a string literal, the
string will be printed in the error message as well.

JaSpa committed:
[stdlib] Make use of `staticReport` in `_precondition`

With this change the compiler is able to generate error messages if it is able
to verify a precodition will fail at runtime.

Only _precondition's inside @_transparent functions are able to fully leverage
this change because else the compiler can't fold the condition into a true/false
value. The only functions at the moment matching these criteria are `...` and
`..<` for which this feature is requested in SR-707.

gribozavr commented on test/1_stdlib/RangeTraps.swift:
```diff
@@ -30,6 +30,10 @@ import ObjectiveC
 
+// Don't let the compiler see a constant which would lead to compile errors
+// instead of the traps
+let zero = [0][0]
```
Please use `getInt()` from `stdlib/private/StdlibUnittest/OpaqueIdentityFunctions.swift` instead.  (And you can actually inline the call instead of declaring a constant.)

JaSpa replied:
Done

JaSpa commented:
@gribozavr I've addressed your review comments and rebased on master.

At the moment if there is a static error because of a failing precondition this leads to an error on the command line like this:

    file.swift:1:10: error: static error: Can't form Range with end < start

Should I remove the *static error:*  prefix, change it to something else, or keep it like this for now?

gribozavr commented:
I think @jrose-apple was working on this project initially, and I would defer to him for further review.

CodaFi commented:
@jrose-apple Chatted about this briefly.  The concern here is that we haven't audited `_precondition` calls for user-facing error messages.  That probably needs to be done before or during static assert work.

jrose-apple commented:
That's *one* concern. Even back in March it had been a while since I looked at this stuff.

DougGregor commented:
Ping?

JaSpa commented:
I'll rebase on master and resolve the conflicts.
Should I add the same logic for detecting static errors to the user facing functions as well?

jrose-apple commented:
We're about to go on a Thanksgiving at Apple, so take your time. I haven't really looked at your implementation either.

Let's hide this behind a flag for now, so that you can write tests for it but we don't have to audit the precondition messages immediately.

dlbuckley commented:
@JaSpa is this still ongoing?

CodaFi commented:
I'm closing this pull request for now.  Please reopen it after rebasing it if you're still interested in making this change @JaSpa.
